exchanges/tolerance.go

```diff
@@ -0,0 +1,364 @@
+				levels[x].MaxPrice)
+		}
+
+		if levels[x].MinAmount >= levels[x].MaxAmount {
+			return fmt.Errorf("%w for %s %s supplied min: %f max: %f",
+				errInvalidAmountLevels,
+				levels[x].Asset,
+				levels[x].Pair,
+				levels[x].MinAmount,
+				levels[x].MaxAmount)
+		}
+
+		t.minPrice = levels[x].MinPrice
+		t.maxPrice = levels[x].MaxPrice
+		t.stepSizePrice = levels[x].StepPrice
+		t.minAmount = levels[x].MinAmount
+		t.maxAmount = levels[x].MaxAmount
+		t.stepSizeAmount = levels[x].StepAmount
+		t.minNotional = levels[x].MinNotional
+		t.multiplierUp = levels[x].MultiplierUp
+		t.multiplierDown = levels[x].MultiplierDown
+		t.averagePriceMins = levels[x].AveragePriceMins
+		t.maxIcebergParts = levels[x].MaxIcebergParts
+		t.marketMinQty = levels[x].MarketMinQty
+		t.marketMaxQty = levels[x].MarketMaxQty
+		t.marketStepSize = levels[x].MarketStepSize
+		t.maxTotalOrders = levels[x].MaxTotalOrders
+		t.maxAlgoOrders = levels[x].MaxAlgoOrders
+	}
+	return nil
+}
+
+// GetTolerance returns the exchange tolerance parameters for a currency
+func (e *ExecutionTolerance) GetTolerance(a asset.Item, cp currency.Pair) (*Tolerance, error) {
+	e.Lock()
+	defer e.Unlock()
+
+	if e.m == nil {
+		return nil, ErrExchangeToleranceNotLoaded
+	}
+
+	assets, ok := e.m[a]
+	if !ok {
+		return nil, errExchangeToleranceAsset
+	}
+
+	pairs, ok := assets[cp.Base]
+	if !ok {
+		return nil, errExchangeToleranceBase
+	}
+
+	t, ok := pairs[cp.Quote]
+	if !ok {
+		return nil, errExchangeToleranceQuote
+	}
+
+	return t, nil
+}
+
+// CheckTolerance checks to see if the price and amount conforms with exchange
+// level order execution tolerances
+func (e *ExecutionTolerance) CheckTolerance(a asset.Item, cp currency.Pair, price, amount float64, marketOrder bool) error {
+	e.Lock()
+	defer e.Unlock()
+
+	if e.m == nil {
+		// No exchange tolerances loaded so we can nil this
+		return nil
+	}
+
+	assets, ok := e.m[a]
+	if !ok {
+		return errCannotValidateAsset
+	}
+
+	pairs, ok := assets[cp.Base]
+	if !ok {
+		return errCannotValidateBaseCurrency
+	}
+
+	t, ok := pairs[cp.Quote]
+	if !ok {
+		return errCannotValidateQuoteCurrency
+	}
+
+	err := t.Conforms(price, amount, marketOrder)
+	if err != nil {
+		return fmt.Errorf("%w for %s %s", err, a, cp)
+	}
+
+	return nil
+}
+
+// Tolerance defines total limit values for an associated currency to be checked
+// before execution on an exchange
+type Tolerance struct {
+	minPrice         float64
+	maxPrice         float64
+	stepSizePrice    float64
+	minAmount        float64
+	maxAmount        float64
+	stepSizeAmount   float64
+	minNotional      float64
+	multiplierUp     float64
+	multiplierDown   float64
+	averagePriceMins int64
+	maxIcebergParts  int64
+	marketMinQty     float64
+	marketMaxQty     float64
+	marketStepSize   float64
+	maxTotalOrders   int64
+	maxAlgoOrders    int64
+	sync.Mutex
+}
+
+// Conforms checks outbound parameters
+func (t *Tolerance) Conforms(price, amount float64, marketOrder bool) error {
```

Comment:
I think it would be more future-proof if instead of `marketOrder bool` the ordertype is just passed in